unic/char/property/src/macros.rs

```diff
@@ -108,62 +147,220 @@ macro_rules! char_property {
-            type Err = ();
-            fn from_str(s: &str) -> Result<Self, Self::Err> {
-                match s {
-                    $(
-                        stringify!($abbr) => Ok($name::$variant),
-                        stringify!($long) => Ok($name::$variant),
-                    )*
-                    $(
-                        str if ::std::ascii::AsciiExt::eq_ignore_ascii_case(str, stringify!($abbr))
-                            => Ok($name::$variant),
-                        str if ::std::ascii::AsciiExt::eq_ignore_ascii_case(str, stringify!($long))
-                            => Ok($name::$variant),
-                    )*
-                    _ => Err(()),
-                }
-            }
+        char_property! {
+            __impl FromStr for $name; $(
+                $abbr => $name::$variant;
+                $long => $name::$variant;
+            )*
         }
 
-        impl $crate::CharProperty for $name {
-            fn prop_abbr_name() -> &'static str { $abbr_name }
-            fn prop_long_name() -> &'static str { $long_name }
-            fn prop_human_name() -> &'static str { $human_name }
+        char_property! {
+            __impl CharProperty for $name;
+            $abbr_name;
+            $long_name;
+            $human_name;
+        }
+
+        char_property! {
+            __impl Display for $name by EnumeratedCharProperty
         }
 
         impl $crate::EnumeratedCharProperty for $name {
             fn all_values() -> &'static [$name] {
-                const VALUES: &[$name] = &[
-                    $($name::$variant,)+
-                ];
+                const VALUES: &[$name] = &[ $($name::$variant,)+ ];
                 VALUES
             }
-
             fn abbr_name(&self) -> &'static str {
-                match *self {
-                    $( $name::$variant => stringify!($abbr), )*
-                }
+                match *self { $( $name::$variant => stringify!($abbr), )* }
             }
-
             fn long_name(&self) -> &'static str {
-                match *self {
-                    $( $name::$variant => stringify!($long), )*
-                }
+                match *self { $( $name::$variant => stringify!($long), )* }
             }
-
             fn human_name(&self) -> &'static str {
-                match *self {
-                    $( $name::$variant => $human, )*
+                match *self { $( $name::$variant => $human, )* }
+            }
+        }
+    };
+
+    // == Binary Property == //
+
+    (
+        $(#[$name_meta:meta])*
+        pub struct $name:ident(bool) {
+            abbr => $abbr_name:expr;
+            long => $long_name:expr;
+            human => $human_name:expr;
+
+            data_table_path => $data_path:expr;
+
+            pub fn $bool_name:ident(&self) -> bool;
+        }
+
+        pub fn $shorthand_name:ident(char) -> bool;
+
+        $(#[$char_trait_meta:meta])*
+        pub trait $char_trait_name:ident for char {
```

Comment:
I wonder if this trait is pulling its weight. you don't use it in the implementation of `CharBidiClass`, and I suspect most, _if not all_ of the UCD crates will want to expose a trait that has fn for all of its exposed properties. That's just much more convenient to use than a trait for _every single property_. Would you rather:

```rust
import unic_ucd_casing::{UnicodeLowercase, UnicodeUppercase, UnicodeTitlecase, ... };
// or
import unic_ucd_casing::UnicodeCasing;
```

(In the casing case, I suspect we want one trait which exposes a version of `is_lowercase` and friends, and works for both `char` and `str` (in the obvious all form, not the has-a form).)